daceml/onnx/op_implementations/pure_implementations.py

```diff
@@ -0,0 +1,617 @@
+
+            state_exp = sdfg_exp.add_state_after(init_state)
+
+            state_exp.add_mapped_tasklet(
+                '_MatMult_',
+                {'__i%d' % i: '0:%s' % s
+                 for i, s in enumerate([I, J, K])}, {
+                     '_a': dace.Memlet.simple("A", ('__i0, __i2')),
+                     '_b': dace.Memlet.simple("B", ('__i2, __i1'))
+                 },
+                '_c = _a * _b', {
+                    '_c':
+                    dace.Memlet.simple(
+                        "Y", '__i0, __i1', wcr_str='lambda x, y: x + y')
+                },
+                external_edges=True)
+            return sdfg_exp
+
+
+@autoregister_params(op="Identity", name="pure")
+class PureIdentity(ONNXForward):
+    @staticmethod
+    def forward_can_be_applied(node: ONNXOp, state: SDFGState,
+                               sdfg: SDFG) -> bool:
+        return True
+
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+
+        node.validate(sdfg, state)
+
+        def prog(input, output):
+            output[:] = dace.elementwise(lambda x: x, input)
+
+        return program_for_node(prog, sdfg, state, node).to_sdfg()
+
+
+@autoregister_params(op="Reciprocal", name="pure")
+class PureReciprocal(ONNXForward):
+    @staticmethod
+    def forward_can_be_applied(node: ONNXOp, state: SDFGState,
+                               sdfg: SDFG) -> bool:
+        return in_desc_with_name(node, state, sdfg, 'X').dtype in [
+            dace.float16, dace.float32, dace.float64
+        ]
+
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+
+        node.validate(sdfg, state)
+
+        def prog(X, Y):
+            Y[:] = dace.elementwise(lambda x: 1 / x, X)
+
+        return program_for_node(prog, sdfg, state, node).to_sdfg()
+
+
+@autoregister_params(op="Tanh", name="pure")
+class PureTanh(ONNXForward):
+    @staticmethod
+    def forward_can_be_applied(node: ONNXOp, state: SDFGState,
+                               sdfg: SDFG) -> bool:
+
+        in_edges = state.in_edges(node)
+        input_dim = len(in_edges[0].data.subset.size())
+        if input_dim == 2:
+            return True
+
+        return False
+
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+
+        node.validate(sdfg, state)
+        in_edges = state.in_edges(node)
+
+        ii = in_edges[0].data.subset.size()[0]
+        jj = in_edges[0].data.subset.size()[1]
+
+        I = str(ii)
+        J = str(jj)
+
+        sdfg_exp = dace.SDFG('tanhExpansion')
+        sdfg_exp.add_array('input', (ii, jj), dace.float32)
+        sdfg_exp.add_array('output', (ii, jj), dace.float32)
+
+        state_exp = sdfg_exp.add_state()
+
+        tmp_out = state_exp.add_transient('tmp_out', (ii, jj), dace.float32)
+
+        task1 = state_exp.add_tasklet(
```

Comment:
oh my, what happened here?

Reply:
Oh yes I forgot; I will try to do elementwise tanh here